src/main/java/htsjdk/variant/variantcontext/JEXLMap.java

```diff
@@ -11,152 +12,188 @@
+        jexl.putAll(map);
+    }
+
+    // -----------------------------------------------------------------------------------------------
+    // Utilities
+    // -----------------------------------------------------------------------------------------------
+
+    /**
+     * Lazily initializes {@link #jexl}, in the sense that all keys in {@code jexlCollection} are associated with {@code null}.
+     * @param jexlCollection
+     */
+    private void lazyInitialize(Collection<JexlVCMatchExp> jexlCollection) {
+        jexl = new HashMap<>();
+        for (final JexlVCMatchExp exp: jexlCollection) {
+            jexl.put(exp, null);
+        }
+    }
+
+    /**
+     * Evaluates a {@link JexlVCMatchExp}'s expression, given the current context (and setup the context if it's {@code null}).
+     * @param exp the {@link JexlVCMatchExp} to evaluate
+     * @throws IllegalArgumentException when {@code exp} is {@code null}
      */
-    private void evaluateExpression(JexlVCMatchExp exp) {
+    private void evaluateExpression(final JexlVCMatchExp exp) {
         // if the context is null, we need to create it to evaluate the JEXL expression
-        if (this.jContext == null) createContext();
+        if (this.jContext == null){
+            createContext();
+        }
+
         try {
             final Boolean value = (Boolean) exp.exp.evaluate(jContext);
             // treat errors as no match
             jexl.put(exp, value == null ? false : value);
-        } catch (Exception e) {
+        } catch (final JexlException e) {
+            // TODO: is this decision the desired result?
             // if exception happens because variable is undefined (i.e. field in expression is not present), evaluate to FALSE
+            jexl.put(exp,false);
+            // todo for a todo (must resolve both in PR review rounds): should I remove the todo and code below?
             // todo - might be safer if we explicitly checked for an exception type, but Apache's API doesn't seem to have that ability
-            if (e.getMessage() != null && e.getMessage().contains("undefined variable"))
-                jexl.put(exp,false);
-            else
-                throw new IllegalArgumentException(String.format("Invalid JEXL expression detected for %s with message %s", exp.name, (e.getMessage() == null ? "no message" : e.getMessage())));
+            // if exception happens because variable is undefined (i.e. field in expression is not present), evaluate to FALSE
+            throw new IllegalArgumentException(String.format("Invalid JEXL expression detected for %s with message %s",
+                    exp.name,
+                        /*(e.getMessage() == null ? */"no message" /*: e.getMessage())*/));
+        }
+    }
+
+    /**
+     * Create the internal JexlContext, only when required.
+     * This code is where new JEXL context variables should get added.
+     */
+    private void createContext() {
```

Comment:
I think this was clearer with three distinct cases instead of collapsing one of them.